File: src/BranchPins.Core/BranchRule.cs
using System;
using System.Text.RegularExpressions;

namespace BranchPins.Core;

/// <summary>
/// Maps matching branch names to a shared pin set or disables synchronization for them.
/// </summary>
public sealed class BranchRule
{
    /// <summary>
    /// Initializes a branch rule from configuration values.
    /// </summary>
    /// <param name="pattern">The case-insensitive branch glob pattern.</param>
    /// <param name="pinSetName">The shared pin-set name for matching branches.</param>
    /// <param name="enabled">Whether matching branches are synchronized.</param>
    public BranchRule(string pattern, string pinSetName, bool enabled)
    {
        Pattern = pattern;
        PinSetName = pinSetName;
        Enabled = enabled;
    }

    public string Pattern { get; }

    public string PinSetName { get; }

    public bool Enabled { get; }

    /// <summary>
    /// Tests a branch name against this rule's case-insensitive <c>*</c> and <c>?</c> glob pattern.
    /// </summary>
    /// <param name="branchName">The branch name to test.</param>
    /// <returns><see langword="true"/> when the branch name matches; otherwise <see langword="false"/>.</returns>
    public bool IsMatch(string branchName)
    {
        return Regex.IsMatch(branchName, "^" + Regex.Escape(Pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$", RegexOptions.IgnoreCase);
    }
}

File: vscode/src/core/branchRule.ts
/** Configures a branch glob, its shared pin-set name, and whether synchronization is enabled. */
export interface BranchRuleConfiguration {
    pattern: string;
    pinSetName: string;
    enabled: boolean;
}

/** Tests a branch name against a rule's case-insensitive `*` and `?` glob pattern. */
export function branchRuleMatches(rule: BranchRuleConfiguration, branchName: string): boolean {
    const escaped = rule.pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*").replace(/\?/g, ".");
    return new RegExp("^" + escaped + "$", "i").test(branchName);
}

/** Finds the longest (most specific) matching configured branch glob. */
export function findMatchingRule(
    rules: readonly BranchRuleConfiguration[],
    branchName: string,
): BranchRuleConfiguration | undefined {
    let best: BranchRuleConfiguration | undefined;
    for (const rule of rules) {
        if (!rule.pattern || rule.pattern.trim().length === 0) {
            continue;
        }

        if (!branchRuleMatches(rule, branchName)) {
            continue;
        }

        if (!best || rule.pattern.length > best.pattern.length) {
            best = rule;
        }
    }

    return best;
}
